src/interactive-window/interactiveWindow.ts

```diff
@@ -461,7 +432,14 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
                 // Clear cached kernel when the selected controller for this document changes
                 if (e.controller.id !== this.currentKernelInfo.controller?.id) {
                     this.disconnectKernel();
-                    this.startKernel(e.controller.controller, e.controller.connection).ignoreErrors();
+                    this.currentKernelInfo = {
+                        controller: e.controller.controller,
+                        metadata: e.controller.connection
+                    };
+                    // don't start the kernel if the IW has only been restored from a previous session
+                    if (this.initialized) {
+                        this.startKernel().ignoreErrors();
```

Comment:
this event is what sets the controller on restore. It is called even when the controller is initially set, and will also take care of the case that the user changes the controller before running any cells